app/src/main/java/nu/yona/app/security/EncryptionUtils.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * <?xml version="1.0" encoding="utf-8"?><!--
+ * ~ Copyright (c) 2018 Stichting Yona Foundation
+ *   ~
+ *   ~ This Source Code Form is subject to the terms of the Mozilla Public
+ *   ~ License, v. 2.0. If a copy of the MPL was not distributed with this
+ *   ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *   -->
+ */
+
+package nu.yona.app.security;
+
+import android.content.Context;
+import android.os.Build;
+
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+
+import nu.yona.app.utils.AppUtils;
+
+
+public class EncryptionUtils
+{
+
+	public static String encrypt(Context context, String token)
+	{
+		SecurityKey securityKey = getSecurityKey(context);
+		return securityKey != null ? securityKey.encrypt(token) : null;
+	}
+
+	public static String decrypt(Context context, String token)
+	{
+		SecurityKey securityKey = getSecurityKey(context);
+		return securityKey != null ? securityKey.decrypt(token) : null;
+	}
+
+	private static SecurityKey getSecurityKey(Context context)
+	{
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
+		{
+			return EncryptionKeyGenerator.generateSecretKey(getKeyStore());
+		}
+		return EncryptionKeyGenerator.generateSecretKeyPreM(context, getKeyStore());
```

Comment:
How does it work when a user upgrades from pre-M to M or later? Does the old key still work? Or do we need separate encryption methods for this?

Reply:
Great catch! I did not consider that scenario. Old encryption should still work if we use it, but app will try to decrypt with latest available encryption based on current OS. 
I will update that and test it again.

Reply:
Finding ways yet to replicate and test the scenario with simulators, as physical devices are not available with required sepcifications.

Reply:
How is this now being addressed?